libsolidity/codegen/CompilerContext.cpp

```diff
@@ -265,31 +265,38 @@ void CompilerContext::appendInlineAssembly(
 	}
 
 	unsigned startStackHeight = stackHeight();
-	auto identifierAccess = [&](
+
+	assembly::ExternalIdentifierAccess identifierAccess;
+	identifierAccess.resolve = [&](
+		assembly::Identifier const& _identifier,
+		assembly::IdentifierContext
+	) {
+		auto it = std::find(_localVariables.begin(), _localVariables.end(), _identifier.name);
+		return it == _localVariables.end() ? size_t(-1) : 1;
+	};
+	identifierAccess.generateCode = [&](
 		assembly::Identifier const& _identifier,
-		eth::Assembly& _assembly,
-		assembly::CodeGenerator::IdentifierContext _context
+		assembly::IdentifierContext _context,
+		eth::Assembly& _assembly
 	) {
 		auto it = std::find(_localVariables.begin(), _localVariables.end(), _identifier.name);
-		if (it == _localVariables.end())
-			return false;
+		solAssert(it != _localVariables.end(), "");
 		unsigned stackDepth = _localVariables.end() - it;
 		int stackDiff = _assembly.deposit() - startStackHeight + stackDepth;
```

Comment:
In your list, everything below "let declared variables" is essentially the same thing. If the assembler cannot find an identifier in its own list, it asks the environment and that can respond as it may please. This responds consists (apart from found / not found) of the stack size and the code to inject for accessing the variable.

When inline assembly is used by the Solidity code generator, no user-defined variables are visible. It is only possible to give names to pre-existing stack slots, everything has exactly one stack slot and read / write operations are the usual DUPi / SWAPi POP.

When inline assembly is used by the programmer using the `assembly { ... }` construct, local solidity variables (which includes function arguments), state variables, local functions and libraries are visible.